### PR TITLE
Neuralized K-Means

### LINKED ISSUE #198: Neuralized K-Means: make k-means amenable to neural network explanations
## Description
K-Means finds some centroids $\mathbf{\mu}_1,\ldots,\mathbf{\mu}_K$ and assigns data points to a cluster as $c = {\rm argmin}_k \lbrace \Vert\boldsymbol{x} - \mathbf{\mu}_k\Vert^2\rbrace$, or in code:

```
c = torch.argmin(torch.cdist(x, centroids)**2)
```

Neither the assignment $c$ nor the distance $\Vert\boldsymbol{x} - \mathbf{\mu}_c\Vert^2$ are really explainable.
The assignment $c$ is not continuous, and the distance is in fact measuring a dissimilarity to the cluster, essentially the opposite of what we want to explain.

## Fixes
[From Clustering to Cluster Explanation via Neural Networks (2022)](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=9817459) present a method based on neuralization. They show that the k-means cluster assignment can be exactly rewritten as a set of piecewise linear functions 
```math
f_c(\boldsymbol{x}) = \min_{k\neq c}\lbrace \boldsymbol{w}_{ck}^\top\boldsymbol{x} + b_{ck}\rbrace
```
and the original k-means cluster assignment can be recovered as $c = {\rm argmax}_k\lbrace f_k(\boldsymbol{x})\rbrace$.

The cluster discriminant $f_c$ is a measure of cluster membership (instead of a dissimilarity) and is also structurally amenable to LRP and friends. It can also be _plugged_ on top of a neural network feature extractor to make deep cluster assignments explainable.

## Additional Information
- There is no _pairwise distance_ layer in PyTorch. Since zennit canonizers apply to `torch.nn.Module` modules, such module would need to be added
- Each discriminant $f_c$ has it's own weight matrix $W\in\mathbb{R}^{(K-1)\times D}$. In principle, a neuralized k-means layer could be written as a layer 
```
torch.nn.Sequential(
    torch.nn.Linear(in_features=D, out_features=K*(K-1)), 
    torch.nn.Unflatten(1, torch.Size([K, K-1]))
)
```
or alternatively via `torch.einsum`, which I think is easier. Either way, a special layer is needed for that.
- The paper suggests a specific LRP rule for the $\min_{k\neq c}\lbrace s_k\rbrace$ to make the explanations continuous. In practice, we can replace it by $\beta^{-1}{\rm LogMeanExp}_k\lbrace \beta\cdot s_k\rbrace$ which approximates the min for $\beta < 0$, but has continuous gradients. Gradient propagation through this layer recovers exactly the LRP rule from the paper, but this also provides explanation continuity for other gradient-based explanation methods. I would suggest to implement such a `LogMeanExpPool` layer. Alternative could be to implement a `MinPool` layer with a custom LRP rule.
- A canonizer could be used to replace the distance/k-means layer by it's neuralized counterpart. Alternatively, a composite might also work, but it seems a bit more complicated (where are the weights computed? how to set LRP rules for the linear/einsum layer?)

### PR DESCRIPTION
**Pull request by jackmcrider**
Hi chr5tphr!

I started an attempt to implement (deep) neuralized k-means (https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=9817459) as more people want to use it and ask for code.

I took the SoftplusCanonizer from the docs as a starting point.

Main changes:
- Add Distance, NeuralizedKMeans and LogMeanExpPool in zennit.layer
- Add KMeansCanonizer in zennit.canonizers

Some things can be optimized:
- add tests (I wrote a few, but then changed something and did not bother to rewrite tests unless I see a chance of this being merged)
- copy.deepcopy in KMeansCanonizer.register probably not ideal
  - alternative: clone distance_module.centroids and import Distance from layer.py; then, create new Distance instance in remove()
- KMeansCanonizer.apply scans for Distance layers, but there can be distance layers in some architectures that don't belong to k-means
- one idea would be add a "contrastive layer" and identify kmeans as a composition of Distance followed by Contrastive similar to MergeBatchNorm:
  - contrastive computes ```(out[:,None,:] - out[None,:,:])[mask].reshape(K,K-1,D)```, cf. line 379-384 in canonizers.py
- several advantages: 
  - output of k-means and neuralized k-means are identical,
  - contrastive layer could also be applied to classifiers to get class-contrastive explanations (cf. Fig. 10.5 in https://iphome.hhi.de/samek/pdf/MonXAI19.pdf)
  - could replace NeuralizedKMeans layer by something more general (difference between squared distances and difference between linear layers are both linear, i.e. a contrastive layer covers both; would still need two separate Canonizers)
- the scanning in KMeansCanonizer.apply does not work if Distance is the first layer (kmeans in input space), one can do ```Sequential(Distance(centroids))``` as a trick, but not ideal
- if I understand correctly, KMeansCanonizer.register is executed each time one calls ```with Gradient(...) as attributor```; could be a bottleneck if number of clusters is large

Closes #198

### CODE REVIEW

File: src/zennit/layer.py [hunk 1]
```diff
@@ -34,3 +34,52 @@ def __init__(self, dim=-1):
     def forward(self, input):
         '''Computes the sum along a dimension.'''
         return torch.sum(input, dim=self.dim)
+
+
+class Distance(torch.nn.Module):
+    '''Compute distance between inputs and centroids.'''
+    def __init__(self, centroids, power=2):
+        super().__init__()
+        self.centroids = torch.nn.Parameter(centroids)
```
Comment (p16i) [inline src/zennit/layer.py:43]: Perhaps, we could add a sentence the docstring describing the expected dimensions of  `centroids`.

**Comment by jackmcrider** [comment]
I just found the contributing guide and converted this to draft for now since I broke every single guideline.

@p16i I clicked somewhere and triggered a review request. Please ignore.

**Comment by jackmcrider** [comment]
I tried to merge everything into one commit, extended documentation and made sure that all checks pass.
Could be reviewed, but there are no functional changes.

**Comment by jackmcrider** [comment]
Checks pass 👍 

It's quite challenging to get reproducible tox results for tutorials (e.g. had to manually fiddle with `metadata.kernelspec.name` in the raw `.ipynb` before committing). Probably a limitation of tox, but could be documented for future contributors. I'm not sure where, maybe in [Contributing#continuous-integration](CONTRIBUTING.md#continuous-integration).

I'm gonna freeze this branch for now, unless something comes up.

File: src/zennit/layer.py [hunk 2]
```diff
@@ -34,3 +34,132 @@ def __init__(self, dim=-1):
...
+        Returns
+        -------
+        :py:obj:`torch.Tensor`
+            shape (N, K, K-1) tensor of k-means discriminants
+        '''
+        x = torch.einsum('nd,kjd->nkj', x, self.weight) + self.bias
+        return x
+
+
+class LogMeanExpPool(torch.nn.Module):
```
Comment (chr5tphr) [inline src/zennit/layer.py:122]: I think this should not be called *Pool*, as Pooling is usually not done over whole axes, but over subsets of the input (e.g., MaxPool2d, AvgPool1d...)

File: src/zennit/types.py [hunk 3]
```diff
@@ -124,3 +126,10 @@ class Activation(metaclass=SubclassMeta):
         torch.nn.modules.activation.Tanhshrink,
         torch.nn.modules.activation.Threshold,
     )
+
+
+class Distance(metaclass=SubclassMeta):
+    '''Abstract base class that describes distance modules.'''
```
Comment (chr5tphr) [inline src/zennit/types.py:132]: I do not think this class is necessary, since this is not used for general distance modules, but rather only for the `PairwiseCentroidDistance` module, as the `centroids` are accessed after an instance-check in the canonizer.
Unless there is another use, this can be removed, and the canonizer can directly use `PairwiseCentroidDistance`.

Or will there be more centroid-based distance layer types?

File: src/zennit/canonizers.py [hunk 4]
```diff
@@ -329,3 +331,100 @@ def register(self):
...
+
+        Parameters
+        ----------
+        root_module : :py:obj:`torch.nn.Module`
+            Root module containing a :py:obj:`Distance` layer with power 2 as a submodule.
+        '''
+        instances = []
+
+        for full_name, module in root_module.named_modules():
+            if isinstance(module, Distance) and module.power == 2:
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:379]: See comment on the typing.
```suggestion
            if isinstance(module, PairwiseCentroidDistance) and module.norm == 2 and not module.root:
```

File: src/zennit/canonizers.py [hunk 5]
```diff
@@ -18,10 +18,12 @@
 '''Functions to produce a canonical form of models fit for LRP'''
 from abc import ABCMeta, abstractmethod
 
+import copy
 import torch
 
 from .core import collect_leaves
-from .types import Linear, BatchNorm, ConvolutionTranspose
+from .types import Linear, BatchNorm, ConvolutionTranspose, Distance
+from .layer import NeuralizedKMeans, LogMeanExpPool
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:26]:
```suggestion
from .types import Linear, BatchNorm, ConvolutionTranspose
from .layer import NeuralizedKMeans, LogMeanExpPool, PairwiseCentroidDistance
```

File: src/zennit/canonizers.py [hunk 6]
```diff
@@ -329,3 +331,100 @@ def register(self):
...
+
+        n_clusters, n_dims = self.distance.centroids.shape
+        mask = ~torch.eye(n_clusters, dtype=bool)
+        weight = 2 * (self.distance.centroids[:, None, :] - self.distance.centroids[None, :, :])
+        weight = weight[mask].reshape(n_clusters, n_clusters - 1, n_dims)
+        norms = torch.norm(self.distance.centroids, dim=-1)
+        bias = (norms[None, :]**2 - norms[:, None]**2)[mask].reshape(n_clusters, n_clusters - 1)
+        setattr(self.parent_module, self.child_name,
+                torch.nn.Sequential(NeuralizedKMeans(weight, bias),
+                                    LogMeanExpPool(self.beta)))
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:423]: I think the use of `add_module` might be better to replace a module, but it should work in both cases.
```suggestion
        self.parent_module.add_module(
            self.child_name,
            torch.nn.Sequential(NeuralizedKMeans(weight, bias), LogMeanExpPool(self.beta))
        )
```

File: src/zennit/canonizers.py [hunk 7]
```diff
@@ -329,3 +331,100 @@ def register(self):
...
+        A :py:obj:`NeuralizedKMeans` layer is created with these weights and biases. The :py:obj:`LogMeanExpPool` layer
+        is created with the beta value supplied to the constructor.
+
+        Parameters
+        ----------
+        distance_module : list of :py:obj:`Distance`
+            Distance layers to replace.
+        '''
+        self.distance = distance_module
+        self.distance_unchanged = copy.deepcopy(self.distance)
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:413]: Since you are not modifying the distance module, I do not see why we would need a copy. It should be enough to only store the reference to the distance module and later re-set the parent's respective attribute.

File: src/zennit/canonizers.py [hunk 8]
```diff
@@ -329,3 +331,100 @@ def register(self):
...
+        '''Apply this canonizer recursively on all applicable modules.
+
+        Iterates over all modules of the root module and applies this canonizer to all :py:obj:`Distance` layers with
+        power 2.
+
+        Parameters
+        ----------
+        root_module : :py:obj:`torch.nn.Module`
+            Root module containing a :py:obj:`Distance` layer with power 2 as a submodule.
+        '''
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:375]: We might ultimately implement a base-class that replaces child modules. This only serves as a reminder.

File: src/zennit/canonizers.py [hunk 9]
```diff
@@ -329,3 +331,100 @@ def register(self):
...
+        self.distance = distance_module
+        self.distance_unchanged = copy.deepcopy(self.distance)
+
+        n_clusters, n_dims = self.distance.centroids.shape
+        mask = ~torch.eye(n_clusters, dtype=bool)
+        weight = 2 * (self.distance.centroids[:, None, :] - self.distance.centroids[None, :, :])
+        weight = weight[mask].reshape(n_clusters, n_clusters - 1, n_dims)
+        norms = torch.norm(self.distance.centroids, dim=-1)
+        bias = (norms[None, :]**2 - norms[:, None]**2)[mask].reshape(n_clusters, n_clusters - 1)
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:420]:
```suggestion
        bias = (norms[None, :] ** 2 - norms[:, None] ** 2)[mask].reshape(n_clusters, n_clusters - 1)
```

**Review by chr5tphr — CHANGES_REQUESTED** [review]
Hey @jackmcrider 

thanks a lot for the contribution!
I have looked at your implementation and left a few comments.

I have not yet looked at the tutorial.

File: src/zennit/layer.py [hunk 10]
```diff
@@ -34,3 +34,132 @@ def __init__(self, dim=-1):
...
+        input : :py:obj:`torch.Tensor`
+            shape (N, D) tensor of points
+
+        Returns
+        -------
+        :py:obj:`torch.Tensor`
+            shape (N, K) tensor of distances
+        '''
+        distance = torch.cdist(input, self.centroids)**self.power
+        return distance
```
[the remark this answers is not in the document]
Comment (jackmcrider) [reply src/zennit/layer.py:78]: This will not work.

`torch.nn.functional.pairwise_distance` can only compute the diagonal of a square distance matrix.

File: src/zennit/layer.py [hunk 11]
```diff
@@ -34,3 +34,132 @@ def __init__(self, dim=-1):
     def forward(self, input):
         '''Computes the sum along a dimension.'''
         return torch.sum(input, dim=self.dim)
+
+
+class Distance(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (jackmcrider) [reply src/zennit/layer.py:39]: Agreed, but collides with what `torch.nn.PairwiseDistance` does (which is not pairwise at all).
We have to trade-off accuracy for consistency.

I'm in favor of `PairwiseCentroidDistance` and we hope that torch changes the name `PairwiseDistance` to `ElementwiseDistance`.

File: src/zennit/layer.py [hunk 12]
```diff
@@ -34,3 +34,132 @@ def __init__(self, dim=-1):
...
+    >>> centroids = torch.randn(10, 2)
+    >>> distance = Distance(centroids)
+    >>> x = torch.randn(100, 2)
+    >>> distance(x)
+
+    '''
+    def __init__(self, centroids, power=2):
+        super().__init__()
+        self.centroids = torch.nn.Parameter(centroids)
+        self.power = power
```
[the remark this answers is not in the document]
Comment (jackmcrider) [reply src/zennit/layer.py:62]: The implementation of `PairwiseDistance` does not prevent the division by zero issue for negative `p` (the supposed role of `eps`, cf. [source](https://github.com/pytorch/pytorch/blob/387556318eaf60c16fdf9d6f28e749e45a3481ec/aten/src/ATen/native/Distance.cpp#L54)), so we would promote legacy behavior.

It's irrelevant for the purpose of neuralized k-means anyways (which only works for `p=2`), so I don't mind.

File: src/zennit/types.py [hunk 13]
```diff
@@ -18,6 +18,8 @@
 '''Type definitions for convenience.'''
 import torch
 
+from .layer import Distance as DistanceLayer
```
[the remark this answers is not in the document]
Comment (jackmcrider) [reply src/zennit/types.py:21]: Depends on the decision [above](#discussion_r1301846800).

File: src/zennit/layer.py (hunk 2 above)
Comment (jackmcrider) [reply src/zennit/layer.py:121]: After some user feedback, I realized that we should not use this layer at all.

Instead, we should implement a `Min` layer and add a custom rule for that layer.
I will prepare a commit.

A torch `MinPool` would be ideal.
But if I understand correctly, custom rules cannot access the pool windows (without calling `unfold`), which makes this very tricky.

File: src/zennit/types.py (hunk 3 above)
Comment (jackmcrider) [reply src/zennit/types.py:132]: Agreed, can be removed. (will do)

File: src/zennit/canonizers.py (hunk 4 above)
Comment (jackmcrider) [reply src/zennit/canonizers.py:379]: Depends on [above](#discussion_r1301846800).

File: src/zennit/layer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (chr5tphr) [reply src/zennit/layer.py:78]: You're right!
Unfortunate naming in Pytorch, let's keep doing `cdist` then!

Then keepdim and eps can also leave!

File: src/zennit/layer.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (chr5tphr) [reply src/zennit/layer.py:62]: Let's not do of `eps` and `keepdim` then.

File: src/zennit/layer.py (hunk 2 above)
Comment (jackmcrider) [reply src/zennit/layer.py]: Indeed, I forgot about that!
I have implemented the backward-hook via convolution now.

**Comment by jackmcrider** [comment]
I have commited a new version with roughly these changes:
- remove `LogMeanExpPool`
- add layers `MinPool1d` and `MinPool2d` (simple inheritance from the PyTorch `MaxPool*` classes)
- add rules `MinTakesMost1d`, `MinTakesMost2d`, `MaxTakesMost1d`, `MaxTakesMost2d`
- change `KMeansCanonizer` to use `MinPool1d` instead of `LogMeanExpPool`
- change tutorial to use `MinTakesMost1d` at the output layer
- apply changes requested in review (rename `Distance` to `PairwiseCentroidDistance`, do `self.parent_module.add_module` instead of `setattr(self.parent_module...`)

I'm not sure if we want four rules for the \*TakesMost\* or if one rule with `mode='max'`/`mode='min'` and some autodetection for the 1d/2d is better.
